client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,246 @@
+//!
+//! A Registry is implemented by defining the following traits:
+//! - [`Subscribe<K>`];
+//! - [`Dispatch<M>`];
+//! - [`Unsubscribe`].
+//!
+//! As a result of subscription `Hub::subscribe` method returns an instance of
+//! [`Receiver<Message,Registry>`]. That can be used as a [`Stream`] to receive the messages.
+//! Upon drop the [`Receiver<Message, Registry>`] shall unregister itself from the `Hub`.
+
+use std::{
+	collections::HashMap,
+	ops::DerefMut,
+	pin::Pin,
+	sync::{Arc, Weak},
+	task::{Context, Poll},
+};
+
+use futures::stream::{FusedStream, Stream};
+use parking_lot::{Mutex, MutexGuard};
+
+use crate::{
+	id_sequence::SeqID,
+	mpsc::{TracingUnboundedReceiver, TracingUnboundedSender},
+};
+
+/// Unsubscribe: unregisters a previously created subscription.
+pub trait Unsubscribe {
+	/// Remove all registrations of the subscriber with ID `subs_id`.
+	fn unsubscribe(&mut self, subs_id: SeqID);
+}
+
+/// Subscribe using a key of type `K`
+pub trait Subscribe<K> {
+	/// Register subscriber with the ID `subs_id` as having interest to the key `K`.
+	fn subscribe(&mut self, subs_key: K, subs_id: SeqID);
+}
+
+/// Dispatch a message of type `M`.
+pub trait Dispatch<M> {
+	/// The type of the that shall be sent through the channel as a result of such dispatch.
+	type Item;
+	/// The type returned by the `dispatch`-method.
+	type Ret;
+
+	/// Dispatch the message of type `M`.
+	///
+	/// The implementation is given an instance of `M` and is supposed to invoke `dispatch` for
+	/// each matching subscriber, with an argument of type `Self::Item` matching that subscriber.
+	///
+	/// Note that this does not have to be of the same type with the item that will be sent through
+	/// to the subscribers. The subscribers will receive a message of type `Self::Item`.
+	fn dispatch<F>(&mut self, message: M, dispatch: F) -> Self::Ret
+	where
+		F: FnMut(&SeqID, Self::Item);
+}
+
+/// A subscription hub.
+///
+/// Does the subscription and dispatch.
+/// The exact subscription and routing behaviour is to be implemented by the Registry (of type `R`).
+/// The Hub under the hood uses the channel defined in `crate::mpsc` module.
+#[derive(Debug)]
+pub struct Hub<M, R> {
+	tracing_key: &'static str,
+	shared: Arc<Mutex<Shared<M, R>>>,
+}
+
+/// The receiving side of the subscription.
+///
+/// The messages are delivered as items of a [`Stream`].
+/// Upon drop this receiver unsubscribes itself from the [`Hub<M, R>`].
+#[derive(Debug)]
+pub struct Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	rx: TracingUnboundedReceiver<M>,
+
+	shared: Weak<Mutex<Shared<M, R>>>,
+	subs_id: SeqID,
+}
+
+#[derive(Debug)]
+struct Shared<M, R> {
+	id_sequence: crate::id_sequence::IDSequence,
+	registry: R,
+	sinks: HashMap<SeqID, TracingUnboundedSender<M>>,
+}
+
+impl<M, R> Hub<M, R>
+where
+	R: Unsubscribe,
+{
+	/// Provide mutable access to the registry (for test purposes).
+	pub fn lock_registry<'a>(&'a self) -> impl DerefMut<Target = R> + 'a {
```

Comment:
If this is only for testing and is not supposed to be used in normal code then *personally* I'd explicitly name it as such e.g. by putting in the name that it's only for testing, something like `lock_registry_FOR_TESTS`. (Unfortunately we don't do that very often so I'm not going to force it; feel free to ignore this if you want.)

Reply:
Or when we only use this internally for tests (not verified), we can guard it as well.

Reply:
I decided to opt for 
```rust
pub fn map_registry_for_tests<MapF, Ret>(&self, map: MapF) -> Ret
	where
		MapF: FnOnce(&R) -> Ret
```
as it is a safer thing to have in a public API.